keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,191 @@
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
+
+Optional: In addition to progress notify, also configure etcd to send regular
+progress events all on watches, every 10-100 milliseconds or so. If this is
+works, it could reduce the need to need to send progress notify requests to
+etcd, since watch cache progress would already kept largely up-to-date
+by the progress events etcd proactively sent.
+
+Optional: For some or all of the etcd progress watch events, also create a
+kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability).
+
+### Risks and Mitigations
+
+This can increase the number of round trips to etcd: One is required to get the
+latest revision, another might be needed to request the progress notify. Both of
+these requests are cheap and scalable, but we need to understand the latency
+impact of waiting for both of them before serving a response back to the
+client. If this turns out to have performance implications, it could be
+partially mitigated by configuring etcd to send regular progress events.
+
+In the worst case, the “wait until fresh” duration is exceeded, and then the
+request must be served from etcd anyway, which will be much higher latency than
+immediately serving from etcd like we do today. If this is an issue, it might be
+possible to partially mitigate it by tuning the “wait until fresh” duration to a
+sufficiently short value.
+
+## Design Details
+
+### Test Plan
+
+We will need to scale and performance test this carefully. We’ll need to measure
+latency, throughput, volume of progress notify requests, and change to volume of
+list requests served from storage instead of from cache.
+
+### Rollout Plan
+
+#### Serving consistent reads from cache
+
+Guard it with the `WatchCacheConsistentReads` feature gate.
+Communicate that clusters administrators should set `--storage-backend` with major.minor version, .e.g. `--storage-backend=etcd3.4` to benefit from features requiring newer etcd versions.
+
+#### Reflector
+
+- Provide a way for reflectors to be configured to use resourceVersion=”” for initial list
+- Upgrade all in-tree reflectors to use resourceVersion=”” (TODO: this will be problematic for cluster that do not have this feature enabled)
```

Comment:
We should evaluate it when we will be ready with the implementation.
But assuming we will hear a lot of feedback that people are not yet using etcd 3.4, if we would expose the option above (ability to configure reflector) via flag or some other configuration option on components, that would allow everyone to configure it.
And we can leave the default still set to "0", and switch it once this will appear to be much safer.

Reply:
Yes, this is the best available approach. I've added it to the doc.